src/tools/genpolicy/rules.rego

```diff
@@ -55,14 +53,178 @@ default AllowRequestsFailingPolicy := false
+  p_env_map := p_container.env_map  
+  allow_env_map(p_env_map, i_env_map)
+
+  i_process := input.base.OCI.Process
+  p_process := p_container.OCI.Process
+
+  allow_args_by_equality(i_process, p_process, i_env_map)
+
+  print("PolicyCreateContainerRequest: true")
+}
+
+allow_args_by_equality(i_process, p_process, i_env_map) {
+  not i_process.Args
+  not p_process.Args
+  print("allow_args_by_equality: no args")
+}
+
+allow_args_by_equality(i_process, p_process, i_env_map) {
+  i_args := i_process.Args
+  p_args := p_process.Args
+  print("allow_args_by_equality: i_args =", i_args, "p_args =", p_args)
+  count(i_args) == count(p_args)
+  every i, i_arg in i_args {
+    print("allow_args_by_equality: i_arg =", i_arg)
+    p_arg_replaced := replace_env_variables(p_args[i], i_env_map)
+    # Double $$ are reduced to a single $, from command field description in https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.24/#container-v1-core
+    p_arg_replaced2 := replace(p_arg_replaced, "$$", "$")
+    print("allow_args_by_equality: p_arg_replaced2 =", p_arg_replaced2)
+    i_arg == p_arg_replaced2
+    print("allow_args_by_equality: true")
+  }
+}
+
+# this function replaces all the environment variables in a string, given a map of environment keys to environment values
+# eg str = "echo $(CLUSTER_ID); echo $(NODE_NAME);"
+# env_map = {"CLUSTER_ID": "abc", "NODE_NAME" : "xyz"}
+# result = "echo abc; echo xyz;"
+replace_env_variables(str, env_map) = result {
+  # make an array of the keys eg ["CLUSTER_ID", "NODE_NAME"]
+  keys := [x | some x in object.keys(env_map)]
+  result := replace_str_rec(str, env_map, keys, count(keys) - 1)
+}
+
+# base case
+replace_str_rec(str, env_map, arr_keys, i) = result {
+  i < 0
+  result = str
+}
+# recursive step
+replace_str_rec(str,env_map, arr_keys, i) = result {
+  i >= 0
+  key := arr_keys[i]
+
+  env_key1 := concat("", ["$(", key, ")"])
+  # replace $(VAR) with value
+  new_str1 := replace(str, env_key1, env_map[key])
+
+  # replace next environment variable
+  result = replace_str_rec(new_str1, env_map, arr_keys, i - 1)
+}
+
+allow_env_map(p_env_map, i_env_map) {
+    every env_key, env_val in i_env_map {
+      print("allow_env: env_key =", env_key, "env_val =", env_val)
+      allow_env_map_entry(env_key, env_val, p_env_map)
+    }
+    print("allow_env_map: true")
+}
+
+# Allow exact match
+allow_env_map_entry(key, i_val, p_env_map) {
+    p_val := p_env_map[key]
+    i_val == p_val
+    print("allow_env_map_entry: true")
+}
+
+# Allow variables that should look like a subdomain name
+allow_env_map_entry(key, i_val, p_env_map) {
+  p_val := p_env_map[key]
+  some allowed in ALLOWED_SUBDOMAIN_NAMES
+  p_val == allowed
+  regex.match(SUBDOMAIN_NAME, i_val)
+  print("allow_env_map_entry2: true")
+}
 
-CreateContainerRequest:= {"ops": ops, "allowed": true} {
+# Allow input env variables that match with a request_defaults regex.
+allow_env_map_entry(key, i_val, p_env_map) {
```

Comment:
Although it's not documented, it appears that first rule typically gets evaluated before the following rules with the same name. Therefore, perhaps this more complex allow_env_map_entry should be moved below the less complex variations of allow_env_map_entry.

Reply:
Testing this, it does seem rules get executed top down. However, it seems to be they all get executed all the time, as opposed to immediately stopping execution once one evaluates to true as one would expect (similar to common condition1 || condition2, where condition2 wouldn't even try to execute if condition1 evaluates to true) So unfortunately putting simpler rules at the top would not make the code more efficient. 

![image](https://github.com/user-attachments/assets/f69d9a79-17dc-443f-8ac3-d27b66259789)